enterprise/cmd/repo-updater/authz/perms_syncer.go

```diff
@@ -0,0 +1,275 @@
+	// for details.
+	// TODO(jchen): Ship the initial design to unblock working on authz providers,
+	// but should revisit the feasibility of using ExternalAccount before final delivery.
+
+	usernames, err := fetcher.FetchRepoPerms(ctx, &repo.ExternalRepo)
+	if err != nil {
+		return errors.Wrap(err, "fetch repository permissions")
+	}
+
+	// Get corresponding internal database IDs
+	// TODO(jchen): Remove the use of dbconn.Global().
+	users, err := db.Users.GetByUsernames(ctx, usernames...)
+	if err != nil {
+		return errors.Wrap(err, "get users by usernames")
+	}
+
+	// Set up set of all usernames that need to be bound to permissions
+	bindUsernamesSet := make(map[string]struct{}, len(usernames))
+	for i := range usernames {
+		bindUsernamesSet[usernames[i]] = struct{}{}
+	}
+
+	// Save permissions to database
+	p := &authz.RepoPermissions{
+		RepoID:  int32(repoID),
+		Perm:    authz.Read, // Note: We currently only support read for repository permissions.
+		UserIDs: roaring.NewBitmap(),
+	}
+
+	for i := range users {
+		// Add existing user to permissions
+		p.UserIDs.Add(uint32(users[i].ID))
+
+		// Remove existing user from set of pending users
+		delete(bindUsernamesSet, users[i].Username)
+	}
+
+	pendingBindUsernames := make([]string, 0, len(bindUsernamesSet))
+	for id := range bindUsernamesSet {
+		pendingBindUsernames = append(pendingBindUsernames, id)
+	}
+
+	txs, err := s.permsStore.Transact(ctx)
+	if err != nil {
+		return errors.Wrap(err, "start transaction")
+	}
+	defer txs.Done(&err)
+
+	if err = txs.SetRepoPermissions(ctx, p); err != nil {
+		return errors.Wrap(err, "set repository permissions")
+	} else if err = txs.SetRepoPendingPermissions(ctx, pendingBindUsernames, p); err != nil {
+		return errors.Wrap(err, "set repository pending permissions")
+	}
+
+	return nil
+}
+
+// syncPerms processes the permissions syncing request and remove the request from
+// the quque once it is done (independent of success or failure).
+func (s PermsSyncer) syncPerms(ctx context.Context, request *syncRequest) {
+	defer s.queue.remove(request.typ, request.id, true)
+
+	var err error
+	switch request.typ {
+	case requestTypeUser:
+		err = s.syncUserPerms(ctx, request.id)
+	case requestTypeRepo:
+		err = s.syncRepoPerms(ctx, api.RepoID(request.id))
+	default:
+		err = fmt.Errorf("unexpected request type: %v", request.typ)
+	}
+
+	if err != nil {
+		log15.Warn("Error syncing permissions", "type", request.typ, "id", request.id, "err", err)
+		return
+	}
+}
+
+// RunPermsSyncer starts running the given syncer in the background.
+func RunPermsSyncer(ctx context.Context, syncer *PermsSyncer) {
+	log15.Debug("started perms syncer")
+	defer log15.Info("stopped perms syncer")
+
+	for {
+		select {
+		case <-syncer.queue.notifyEnqueue:
+		case <-ctx.Done():
+			return
+		}
+
+		request := syncer.queue.acquireNext()
+		if request == nil {
+			// No waiting request is in the queue
+			continue
+		}
+
+		syncer.syncPerms(ctx, request)
+	}
```

Comment:
your queue can get stuck, even if there are permissions to sync. For example say you enqueue 3 items onto the queue while on line 273. This will collapse into a single notify on `notifyEnqueue`. You will then only do acquireNext once and wait until someone enqueues again.

Suggestion is to drain the queue here. Not sure on implications on rate limiting/pacing/etc:

```suggestion
    notifyDequed := make(chan struct {}, 1)
	for {
		select {
		case <-notifyDequed:
		case <-syncer.queue.notifyEnqueue:
		case <-ctx.Done():
			return
		}

		request := syncer.queue.acquireNext()
		if request == nil {
			// No waiting request is in the queue
			continue
		}
        notify(notifyDequeued)

		syncer.syncPerms(ctx, request)
	}
```

Reply:
Oh no... I thought I verified it, apparently I didn't do it correctly.

You're correct: https://play.golang.org/p/AW9i9xLvON3